src/mscorlib/shared/System/Number.NumberBuffer.cs

```diff
@@ -0,0 +1,28 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+
+namespace System
+{
+    internal static partial class Number
+    {
+        // Constants used by number parsing
+        private const Int32 NumberMaxDigits = 32; // needs to == NUMBER_MAXDIGITS in coreclr's src/classlibnative/bcltype/number.h
+
+        [StructLayout(LayoutKind.Sequential)]
+        internal unsafe struct NumberBuffer // needs to match layout of NUMBER in coreclr's src/classlibnative/bcltype/number.h
```

Comment:
Declare as "ref struct" so C# can now actually enforce the "stack only" nature of this struct?

Reply:
I hadn't because then I could do Unsafe.AsPointer on ref this (since ref structs can't be used as generic arguments), but I can expand the fixed buffer into individual fields, similar to how you did in corefx.

Reply:
The real problem is that C# has unnecessary limitations on what you can do with fixed buffers. You would want to write this: 
```
public char* digits => (char*)Unsafe.AsPointer(ref _digits[0]);
```
but it won't compile for no good reason.

If you want to save yourself from having 50 padding fields, you can workaround it by doing this:

```
ref struct NumberBuffer
{
   [StructLayout(LayoutKind.Sequential, Size = 50*sizeof(char))]
   struct Digits { }

   private Digits _digits;

   public char* digits => (char*)Unsafe.AsPointer(ref _digits);
}
```

It will generate pretty much the same IL and metadata as what you get for the fixed buffer, except that it will not trigger the unnecessary limitations.

cc @VSadov What it would take to fix the unnecessary limitations on what you can do with fixed buffers?

Reply:
btw, I think I tried the nested ExplicitLayout trick in corefx too but it also had the same bad debugging experience - VS somehow wants to see actual named fields or it won't give you an accurate view of the data.

Reply:
> If you want to save yourself from having 50 padding fields, you can workaround it by doing this

I'll do that, thanks.

Reply:
Hmm.  The current code in coreclr does this:
https://github.com/dotnet/coreclr/blob/fc8bd03dd1dbeab9b7e51e8ecdce3d3c6d9bdfcc/src/mscorlib/src/System/Number.cs#L335
That's 114 on 32-bit and 122 on 64-bit.  But that doesn't seem to account for any padding.  The coreclr code has:
https://github.com/dotnet/coreclr/blame/fc8bd03dd1dbeab9b7e51e8ecdce3d3c6d9bdfcc/src/classlibnative/bcltype/number.h#L22-L29
Are we using pack==1 in native?  Otherwise, it would seem the native size would actually be 120 on 32-bit and 128 on 64-bit.

Reply:
>Are we using pack==1 in native?

I think that's what the "include <pshpack1.h>" is supposed to do.

Reply:
> I think that's what the "include <pshpack1.h>" is supposed to do.

Ok, good, I'll do the same for the managed struct.